internal/hcl/modules/registry.go

```diff
@@ -81,6 +85,53 @@ func (d Disco) ModuleLocation(source string) (RegistryURL, error) {
 	return r, nil
 }
 
+func (d Disco) DownloadLocation(moduleURL RegistryURL, version string) (string, error) {
+	serviceURL, err := d.disco.DiscoverServiceURL(svchost.Hostname(moduleURL.Host), moduleServiceID)
+	if err != nil {
+		return "", fmt.Errorf("unable to discover registry service using host %s %w", moduleURL.Host, err)
+	}
+
+	var u *url.URL
+	if version == "" {
+		u, err = url.Parse(fmt.Sprintf("%s/download", moduleURL.Location))
+	} else {
+		u, err = url.Parse(fmt.Sprintf("%s/%s/download", moduleURL.Location, version))
+	}
+	if err != nil {
+		return "", err
+	}
+
+	downloadURL := serviceURL.ResolveReference(u)
+
+	log.Debugf("Looking up download URL for module %s from registry URL %s", moduleURL.RawSource, downloadURL)
+
+	httpClient := &http.Client{}
+	req, _ := http.NewRequest("GET", downloadURL.String(), nil)
+	moduleURL.Credentials.PrepareRequest(req)
+	resp, err := httpClient.Do(req)
+
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	location := resp.Header.Get("X-Terraform-Get")
+	if location == "" {
+		return "", errors.New("download URL has no X-Terraform-Get header")
+	}
+
+	if strings.HasPrefix(location, "/") || strings.HasPrefix(location, "./") || strings.HasPrefix(location, "../") {
```

Comment:
I think this suggestion is probably okay because I can't think of a case where a location would start with ` .` and not be relative, but it's different to the logic both Terraform CLI and Terragrunt CLI have.